core/drivers/imx_sc_api.c

```diff
@@ -0,0 +1,445 @@
+	union {
+		uint32_t u32[SC_RPC_MAX_MSG - 1];
+		uint16_t u16[(SC_RPC_MAX_MSG - 1) * 2];
+		uint8_t u8[(SC_RPC_MAX_MSG - 1) * 4];
+	} data;
+};
+
+static struct mutex scu_mu_mutex = MUTEX_INITIALIZER;
+static vaddr_t secure_ipc_addr;
+
+register_phys_mem(MEM_AREA_IO_SEC, SC_IPC_BASE_SECURE, SC_IPC_SIZE);
+
+/*
+ * Read a message from an IPC channel
+ *
+ * @ipc	IPC channel
+ * @msg	Received message
+ */
+static TEE_Result sc_ipc_read(vaddr_t ipc, struct sc_rpc_msg *msg)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int count = 0;
+
+	if (!msg) {
+		EMSG("msg is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	assert(ipc);
+
+	res = mu_receive_msg(ipc, 0, (uint32_t *)msg);
+	if (res)
+		return res;
+
+	/* Check the size of the message to receive */
+	if (msg->header.size > SC_RPC_MAX_MSG) {
+		EMSG("Size of the message is > than SC_RPC_MAX_MSG");
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	for (count = 1; count < msg->header.size; count++) {
+		res = mu_receive_msg(ipc, count % MU_NB_RR,
+				     &msg->data.u32[count - 1]);
+		if (res)
+			return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Write a message to an IPC channel
+ *
+ * @ipc	IPC channel
+ * @msg	Send message pointer
+ */
+static TEE_Result sc_ipc_write(vaddr_t ipc, struct sc_rpc_msg *msg)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int count = 0;
+
+	if (!msg) {
+		EMSG("msg is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (msg->header.size > SC_RPC_MAX_MSG) {
+		EMSG("msg->size is > than SC_RPC_MAX_MSG");
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	assert(ipc);
+
+	res = mu_send_msg(ipc, 0, *(uint32_t *)msg);
+	if (res)
+		return res;
+
+	for (count = 1; count < msg->header.size; count++) {
+		res = mu_send_msg(ipc, count % MU_NB_TR,
+				  msg->data.u32[count - 1]);
+		if (res)
+			return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Send an RPC message over the secure world IPC channel
+ *
+ * @msg		Message to send. This pointer will also return the answer
+ *		message if expected.
+ * @wait_resp	Set to true if an answer is expected.
+ */
+static TEE_Result sc_call_rpc(struct sc_rpc_msg *msg, bool wait_resp)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	mutex_lock(&scu_mu_mutex);
+
+	res = sc_ipc_write(secure_ipc_addr, msg);
+
+	if (res == TEE_SUCCESS && wait_resp)
+		res = sc_ipc_read(secure_ipc_addr, msg);
+
+	mutex_unlock(&scu_mu_mutex);
+
+	return res;
+}
+
+/*
+ * Get the partition ID of secure world
+ *
+ * @partition Partition ID
+ */
+static TEE_Result sc_rm_get_partition(uint8_t *partition)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	enum sc_error err = SC_ERR_LAST;
+	struct sc_rpc_msg msg = {
+		.header.version = SC_RPC_VERSION,
+		.header.size = 1,
+		.header.svc = SC_RPC_SVC_RM,
+		.header.func = SC_RM_FUNC_GET_PARTITION,
+	};
+
+	res = sc_call_rpc(&msg, true);
+	if (res != TEE_SUCCESS) {
+		EMSG("Communication error");
+		return res;
+	}
+
+	err = msg.header.func;
+	if (err != SC_ERR_NONE) {
+		EMSG("Unable to get partition ID, sc_error: %" PRId32,
```

Comment:
I think `%d` better applies to enumerated type (not specifically 32bit).

Reply:
Ok